### DISCUSSION TITLE
Evergreen Problem: Chess Difficulty

### DISCUSSION THREAD

**Discussion by Wingdinggaster656** [Monika-After-Story/MonikaModDev#9118] [general]
See #9117
I've noticed that some MAS players who come to Github don't read PRs(as for they are not planning to contribute, they think they probably won't understand those codes) Therefore, I made this discussion. I hope this will not cause any trouble to Devs.

The following is what I said in my PR.

>One thing I might want to tell you is how `Skill Level` work.
>I'm going to make a not so polite assumption here. Forgive me if I'm wrong: Based on your previous discussion of chess systems, it seems that you guys know very little about the "UCI Protocol" which is used by chess engines.
>
>`Skill Level` the parameter, well, let me just paste the Stockfish's document here:
>
>>  * #### Skill Level
>>    Lower the Skill Level in order to make Stockfish play weaker (see also UCI_LimitStrength).
>>   Internally, MultiPV is enabled, and with a certain probability depending on the Skill Level a
>>    weaker move will be played.
>
>Do you understand what it means?
>That said, the `Skill Level` is essentially giving a possibility -- and since it's just a possibility, it's entirely possible for a player to play a game without triggering a weaker move at all(though this chance is indeed very low).
>This is why players often report that "Monika's moves were erratic, sometimes ridiculously strong and sometimes incredibly weak".
>
>The difficulty controlled by `UCI_Elo` -- as far as I can see -- doesn't have this problem.
>But what's the problem now? Stockfish only supports the ELO range from 1350 to 2850.
>
>What I did here is write an `if-else` chain to simulate `Skill Level` below 1350 -- but the problem with `Skill Level`? I have mentioned it.
>And this, would cause Monika to have an unnatural change of style around 1350 points.
>
>One suggestion I have for this is that when the player first reaches 1350, Monika somehow expresses that her style of play is going to change.
>For example, "Seeing your victories these days, I thought you might be interested in learning chess. I will play you more seriously in the future than I did before." Then fixing the lower limit at 1350.
>
>Another idea, of course, would be to ditch the range below 1350 altogether and set the floor at 1350 -- that's not very feasible either. Because even if 1350 is easy enough to reach, but many MAS players don't want to learn it.
>
>I leave the question to you to discuss. If you have any questions about chess engines, feel free to ask me.

Here are some materials that you might consider referring to:
>ELO 800 - A beginner. Anyone who knows chess rules well would reach this.
>ELO 1200 - A "better beginner". No longer than a month, people can reach this, as long as they really studied. By "studied", I didn't mean to train it all day, just 1~2 hours per day would be enough.
>ELO 1600 - Intermediate. Well, reaching this is something you can be proud of. This level still requires no talent, but many people don't have the patience to learn it.
>ELO 2000 - Entry level of the Master world. If you don't have talent, that's pretty much the limit. By arriving here, you are already at the top of your game among amateur players.
>ELO 2500 - Well, you are really good -- You are a Grandmaster now! It takes talent and hard practice to get to this level. Unfortunately, you are only in the middle of the pack among professional players.
>ELO 2800 - You're the top three in the world now. That is to say, you are either Magnus, or Ding Liren, or Alireza Firouzja.
>ELO 2900 - Congratulations on breaking the current world record.
>ELO 3000 - That should be the limit of the human brain. You don't have to try any harder, Albert Einstein II.
>
>(And then, in my setup, Monika's maximum ELO score is 2500, although Stockfish supports 2850.)

**Comment by Justformas** [comment]
> I've noticed that some MAS players who come to Github don't read PRs(as for they are not planning to contribute, they think they probably won't understand those codes) Therefore, I made this discussion. I hope this will not cause any trouble to Devs.

Yeah it's something like that for me right now, although I wouldn't consider myself an average player of MAS or even many games in general, but that's just the way I am (I can get into that more if you want; it might explain some things). I might try to read and understand code, but I don't think I'm ready for making code yet. So, not "contribute" in that sense or at least not yet. But hey, me getting you to work on this project again and giving you feedback is still a win to me.

Anyway what exactly is an Evergreen Problem? I don't think I've heard of that specifically before. Is it related to "the grass is always greener on the other side", or in other words, there are benefits to both things?

I think both things can be implemented like your plan was so the grass can be greener on both sides. So when you say:

> One suggestion I have for this is that when the player first reaches 1350, Monika somehow expresses that her style of play is going to change.
> For example, "Seeing your victories these days, I thought you might be interested in learning chess. I will play you more seriously in the future than I did before." Then fixing the lower limit at 1350.

This seems a bit different than what you were saying on planning on doing before, or at least I didn't see you specify this. My answer is even if the player surpasses 1350, the floor should not be permanently changed to 1350, because then they wouldn't be able to go back. Like I was saying that is (usually) significantly harder than Monika's lower levels right now. BUT similar to what you said about Monika mentioning something when the player passes 1350 Elo, likewise if their Elo score drops below this point of 1350, Monika could not only say something to acknowledge something like this, but could even ask a question to the player:

"Am I playing too hard/seriously for you? I could play less seriously for you. Is that what you want?" Then the player could choose by saying yes or no, and she could give an appropriate response. 

I think that's a good way of handing this, meaning, if difficulty is not going to automatically drop when the player drops below 1350, then ask the question to the player if that's what they want. If you think someone might not hover around or drop below 1350 again, well I think that is possible, and they might want to have an easier time anyway. Remember what Multimokia said, not everyone wants to play  chess with the explicit goal of getting better, but rather to play a game with their partner. I don't think I could have said it better. They may want to have a casual/fun/easier time with Monika, even if they have surpassed 1350 Elo at some point.

Difficulty levels shouldn't become permanently locked or disabled anyway, even if that means that Monika's playstyle has to change between them. Maybe it would be better if she could play more humanlike/consistent below 1350 Elo, but let's take what we can get.

**Reply by Wingdinggaster656** [reply]
> > I've noticed that some MAS players who come to Github don't read PRs(as for they are not planning to contribute, they think they probably won't understand those codes) Therefore, I made this discussion. I hope this will not cause any trouble to Devs.
> 
> Yeah it's something like that for me right now, although I wouldn't consider myself an average player of MAS or even many games in general, but that's just the way I am (I can get into that more if you want; it might explain some things). I might try to read and understand code, but I don't think I'm ready for making code yet. So, not "contribute" in that sense or at least not yet. But hey, me getting you to work on this project again and giving you feedback is still a win to me.
> 
> Anyway what exactly is an Evergreen Problem? I don't think I've heard of that specifically before. Is it related to "the grass is always greener on the other side", or in other words, there are benefits to both things?
> 
> I think both things can be implemented like your plan was so the grass can be greener on both sides. So when you say:
> 
> > One suggestion I have for this is that when the player first reaches 1350, Monika somehow expresses that her style of play is going to change.
> > For example, "Seeing your victories these days, I thought you might be interested in learning chess. I will play you more seriously in the future than I did before." Then fixing the lower limit at 1350.
> 
> This seems a bit different than what you were saying on planning on doing before, or at least I didn't see you specify this. My answer is even if the player surpasses 1350, the floor should not be permanently changed to 1350, because then they wouldn't be able to go back. Like I was saying that is (usually) significantly harder than Monika's lower levels right now. BUT similar to what you said about Monika mentioning something when the player passes 1350 Elo, likewise if their Elo score drops below this point of 1350, Monika could not only say something to acknowledge something like this, but could even ask a question to the player:
> 
> "Am I playing too hard/seriously for you? I could play less seriously for you. Is that what you want?" Then the player could choose by saying yes or no, and she could give an appropriate response.
> 
> I think that's a good way of handing this, meaning, if difficulty is not going to automatically drop when the player drops below 1350, then ask the question to the player if that's what they want. If you think someone might not hover around or drop below 1350 again, well I think that is possible, and they might want to have an easier time anyway. Remember what Multimokia said, not everyone wants to play chess with the explicit goal of getting, but rather to play a game with their partner. I don't think I could have said it better. They may want to have a casual/fun/easier time with Monika, even if they have surpassed 1350 Elo at some point.
> 
> Difficulty levels shouldn't become permanently locked or disabled anyway, even if that means that Monika's playstyle has to change between them. Maybe it would be better if she could play more humanlike/consistent below 1350 Elo, but let's take what we can get.

`Evergreen` literally means a plant (pine tree, for example) that does not shed its leaves all year round, so that it remains green all year round.
`Evergreen Problem` is a slightly poetic use of the word, meaning "the Problem persists over years", and chess difficulty is one evergreen problem: it has been studied for years, but there hasn't been much progress over the years.

As for your suggestion, I will certainly consider it, but I am also going to listen to others. We should leave this discussion here for a while to see what everyone has to say.

**Reply by Justformas** [reply]
> `Evergreen` literally means a plant (pine tree, for example) that does not shed its leaves all year round, so that it remains green all year round. `Evergreen Problem` is a slightly poetic use of the word, meaning "the Problem persists over years", and chess difficulty is one evergreen problem: it has been studied for years, but there hasn't been much progress over the years.

I see about the Evergreen Problem now. I don't know if you noticed but I mentioned an old chess game back in the other discussion. It's likely chess games back then worked differently compared to chess engines now, but the one I mentioned seemed to get very easy, probably comparable to 200 Elo (which you say is possible here). Not sure how hard the highest level in that game would be considered in Elo standards though. Anyway maybe it's difficult to combine older types of playstyles like that and newer styles into one engine. Or for that matter, to use more than one chess engine in the same game/application. I find the concept interesting, but I'm not suggesting to do any of that. You say it's possible to get it down to 200 Elo anyway.

> As for your suggestion, I will certainly consider it, but I am also going to listen to others. We should leave this discussion here for a while to see what everyone has to say.

So here's another thing: Let's say I could (either now or one day) make it past your skill level 6. Personally I might be discouraged from even going past it in the first place. knowing that I can't go back to difficulty levels I might enjoy more. If I have to even worry about that then I don't think that's a good thing. I have another idea, but I think it might be more complicated and not necessary; have 2 different chess modes for this. 

I can't see most of the developers thinking that permanently raising the difficulty floor to 1350 Elo once you pass that point is a good idea either, but I could be wrong. If you are a serious chess player and/or are truly dedicated to improving, then at some point you should pass 1350 and get higher anyway, right? Those people shouldn't have to worry about going below 1350 Elo anyway. Unless they get rusty, but then, maybe they'd want difficulty to go below 1350 again anyway.

**Reply by Wingdinggaster656** [reply]
(In fact, anyone who has actually played chess knows that it's really no fun and boring to play against lower-level players, and I'm assuming that everyone with 1350 points is basically someone interested in chess.)

You're right, so you're saying we ignore the unnatural change in section 1350? Or some other topics?

**Reply by Justformas** [reply]
> (In fact, anyone who has actually played chess knows that it's really no fun and boring to play against lower-level players, and I'm assuming that everyone with 1350 points is basically someone interested in chess.)

I am probably below 1350 Elo then. I thought I was anyway based on playing against 1300 on chess.com. However I think 200/250 is pretty easy and I can even find enjoyment in playing against that. I think I'd have to improve quite a bit to consider anything below 1350 boring. Even then I'm still skeptical about that, because due to what I consider fun or what type of experience I might be looking for, I'm not convinced that I couldn't still get some type of enjoyment out of it.

> You're right, so you're saying we ignore the unnatural change in section 1350? Or some other topics?

That's exactly what I'm saying, to either ignore the unnatural change or accept it. I think your implementation of levels lower than 1350 is going to be better than what we have in the mod right now anyway. Besides if we can accept a change in playstyle for difficulty going up, then we should be able to accept it for difficulty going down too. Even if average MAS players notice this change in playstyle, I think most could accept it. Most average MAS players may not pass 1350 anyway.

**Comment by Justformas** [comment]
Reposting my response to your last comment in the PR here then, since it's more appropriate to talk about here.

I think even before your "edit twice" I could see why an experienced/serious chess player could take a good guess at ELO in a match. So actually now, "I want to play a game". I don't remember if you told me specifically, but based on everything I remember you telling me, I want to see if I can guess a ballpark figure on your ELO. I think you aren't lower than 2100 or not by much, but I think you are likely higher, maybe even around 2500. I would be a little surprised if you are actually much higher than that, but not that surprised. So are you somewhere in the range of 2100 to 2500?

Oh and I had a question about something else you said:

>I decided to play 50 bullet chess game with Monika at 400 ELO, probably cost me a few hours.

I think you mean here that you played 50 games while more or less intentionally limiting your time to within 3 minutes per game, right? But are you actually planning on adding time limit modes? Even if you do, then there should still be the unlimited time mode for the player too.

**Reply by briaryoung** [reply]
The move is much appreciated!

**Reply by Wingdinggaster656** [reply]
No plan to make time limit mode, because Monika is, actually, sadly, just a robot .
Would you dare time trial a robot? I wouldn't dare, and neither would anyone I know.
If I had Monika simulate a real person (say very fast in the opening stage and very slow in the middle game), it would have taken too much effort and not been necessary.

As for wanting to match with me...... I will set up a new chess.com account for you. I will post the new account here later when I am finished.
I did it because I didn't want to reveal who I really was.

Also I may be planning to delete my communication with you in that PR, so that PR is much neat than now.

EDIT: Alright I created my new account now. Name is `WDG656`, just created today. We can also continue our conversation about "chess things" not "MAS things" there.

**Reply by Justformas** [reply]
Wait I didn't mean to play a chess match with you. I highly doubt I'm skilled enough for that, but I do feel like I may have gotten slightly better since this whole thing started. But I made an account there and messaged you; I can explain what I meant and talk more there.

> No plan to make time limit mode, because Monika is, actually, sadly, just a robot .
> Would you dare time trial a robot? I wouldn't dare, and neither would anyone I know.
> If I had Monika simulate a real person (say very fast in the opening stage and very slow in the middle game), it would have taken too much effort and not been necessary.

I wasn't suggesting to, I was just curious if you were or were just referring to your fast matches against her. So you were just referring to that. Also I didn't mean it like giving her more thinking time or making the player wait more either, just a time limit which probably only the player would ever run up against as an obstacle. Either way, yeah I don't think I'd want to do that.

**Reply by Justformas** [reply]
> Also I may be planning to delete my communication with you in that PR, so that PR is much neat than now.

By the way go ahead and delete them if you want. Then I might delete my messages there too. I had already marked my last response there as off topic anyway.

Also not sure if you saw my edit of my previous response here, but I did make an account on chess.com. I didn't want to play a match against you but did send you a message and want to talk more about chess stuff there, though not right this moment.

**Reply by Wingdinggaster656** [reply]
> > Also I may be planning to delete my communication with you in that PR, so that PR is much neat than now.
> 
> By the way go ahead and delete them if you want. Then I might delete my messages there too. I had already marked my last response there as off topic anyway.
> 
> Also not sure if you saw my edit of my previous response here, but I did make an account on chess.com. I didn't want to play a match against you but did send you a message and want to talk more about chess stuff there, though not right this moment.

See it now. I had sent a friend request to you.

**Comment by Justformas** [comment]
Hey @multimokia I'll just ask you myself. What exactly did you (or you guys) do to lower the chess difficulty the first time a few years ago? Because it doesn't seem like it was just lowering skill level, default skill level and depth for "sublevels". Stockfish still plays excellently if those are the only things that get changed, even at skill level 0 depth 1. Monika seems to somewhat frequently make worse moves than Stockfish would on those settings. In fact, she may even play worse than if MultiPV is set to 4 or even 8 on her lowest levels. So what else did you do if anything? And don't be afraid to get technical.

**Reply by Wingdinggaster656** [reply]
Here is just to tell you the results of my research:
The changes they made are only the `Skill Level` and `depth` (in fact, they also set `ponder`, but the variable `ponder` is meaningless here, you can learn its specific meaning if you are interested in it).

If there was any potential change, it was to the Stockfish code. But if that were the case, they would have released the code, otherwise they would be in serious violation of the GNU protocol.

**Reply by Wingdinggaster656** [reply]
And I don't understand why you said the current version was easier. Based on the tests you and I did earlier, isn't my new version easier?

(Even so, the old version was strange...... I have some suspicion that Stockfish was changed source code, just not by the current Devs, so they don't know)

**Reply by Justformas** [reply]
A few of the tests with Stockfish at MultiPV 8 against 400 ELO seemed to have Stockfish appear easier, though Stockfish did still win those games. However when I tried playing against your new system it seemed harder, even at 200 ELO. I had also tried playing against MultiPV 4 which seemed harder than MultiPV 8.

I think there used to be a "bug" where Stockfish could drop below skill level 0, which could cause it to loop around to skill level 20. There might also be a case where skill level could get locked at 20 like if you cheat enough. People were saying even a Stockfish at full strength had a hard time beating her, but that might be at least partly why people thought it seemed harder in the past.

**Comment by Justformas** [comment]
I'm going to reply to this here for now in case I get too off-topic for the PR or something.

>Note: Well I eventually came back later to test this PR again.
I have observed something strange: Monika seems to have a difficulty shift by mistake when the ELO score is extremely low (below 400), resulting in Monika's actual strength being close to 2000.

>I don't have time to look into the cause for now. I'll come back in a week to check this. 

So I was wondering if low ELO could cause that to happen, like because of an overflow or something. Both times that happened to me it was below 400. But really you think it's that high? As in around 2000 (and for that matter, in MAS right now you think Monika plays at around 1200 to 1600)? Now I want to test it at 400 or a little higher just to see how it feels, because I don't think I did before.

**Reply by Wingdinggaster656** [reply]
2000 isn't an exact number, it's just a rough feeling.
But anyways it's not normal. I should and will check it. Just not now.

**Reply by Justformas** [reply]
> 2000 isn't an exact number, it's just a rough feeling. But anyways it's not normal. I should and will check it. Just not now.

I'm guessing this might be separate from the other thing/problem you mentioned to me before, right? Because I tried testing it at 400, just over 500 and just over 600, and it doesn't seem any easier than 200 or 300. In fact, it might even be harder. It seems to be about as hard as where I'm at now in MAS's current chess difficulty system, if not harder. Basically it's way harder than 400 - 600 ELO on chess.com.

Edit: I sent you a message on chess.com about another test I did, which might be useful to know.

**Comment by ThePotatoGuy** [comment]
ok ill start actually as a comment here but this is in ref to the PR. currently the changes are small enough that I dont think its a big issue to do before py3. that being said, I think we should look at some improvements:

1. fairy stockfish sounds like the planned engine switch? at least reading the github for that engine, seems like we could open up to tons of chess variants and maybe other games too. I like the potential options and the binary looks tiny. the PR is **not** using that engine though, right, can you update the PR to it? 
     1. also - by this I mean the released fairy stockfish, do not make your own compiled version. 
2. I think changing the how fast the difficulty changes based on *something* is useful for quickly scaling moni to the player. I'll leave it to you to decide if that should be win/loss streaks or win% or something else.
3. you need to have some methodology for converting the player's current difficulty setting to elo. theres 1 comment where I mentioned asking the player for their elo, we should still do that as well but have an option to convert it for the player if they don't know their elo.

**Reply by Wingdinggaster656** [reply]
> can you update the PR to it?

So we are almost inevitable to replace the engine, though we have a few options on this:
1. Stockfish 17
2. Fairy Stockfish
3. Our own Stockfish

Among them, Stockfish 17 is the most stable, while Fairy Stockfish, as you said, is quite scalable. We could provide a large number of variants for players to play with only a small amount of code in the future if we want. The only problem is that the scalability it can provide is still somewhat limited after all. Monika's playing style may still be somewhat non-human, which is a problem that would be difficult to solve without using our own version.

Personally I would say that we pick Fairy Stockfish for now, but still consider using my recompiled version in the future.

> I'll leave it to you to decide if that should be win/loss streaks or win% or something else.

Well isn't that what a Elo system does? Countless online games around the world use this algorithm to calculate player strength, and we just need to emulate them. Furthermore, we can just do something like chess.com, which has a fairly mature modified Elo algorithm specifically designed for chess players.

> you need to have some methodology for converting the player's current difficulty setting to elo

Well this may not make much sense...the current difficulty system can be said to be somewhat disastrous. But if you want, then I think I can.

My idea is that Monika would ask some questions for players. They will not be questions like 'what is your Elo value', but more like  'how long have you been playing chess' and 'are you the kind of person who often plays chess with friends and family?'. Then Monika invite them to play two or three games and we calculate the player's accurate Elo based on these games. 

I think this should be pretty fine for anyone?

**Reply by ThePotatoGuy** [reply]
>Well isn't that what a Elo system does?

your PR currently has static difficulty adjustments though (`_increment_chess_difficulty` and `_decrement_chess_difficulty` are only called with no params so modifier is always 1). so if you add something to move the scale up/down faster based on whatever metric then sure

>My idea is that Monika would ask some questions for players.

ok I think this is ok
